pkg/controller/deployment/sync.go

```diff
@@ -318,35 +318,38 @@ func (dc *DeploymentController) getNewReplicaSet(d *extensions.Deployment, rsLis
 	createdRS, err := dc.client.ExtensionsV1beta1().ReplicaSets(d.Namespace).Create(&newRS)
 	switch {
 	// We may end up hitting this due to a slow cache or a fast resync of the Deployment.
-	// Fetch a copy of the ReplicaSet. If its PodTemplateSpec is semantically deep equal
-	// with the PodTemplateSpec of the Deployment, then that is our new ReplicaSet. Otherwise,
-	// this is a hash collision and we need to increment the collisionCount field in the
-	// status of the Deployment and try the creation again.
+	// Fetch a copy of the ReplicaSet. If the Deployment owns the ReplicaSet and the
+	// ReplicaSet's PodTemplateSpec matches that of the Deployment, it is our new ReplicaSet.
+	// Otherwise, this is a hash collision and we need to increment the collisionCount field
+	// in the status of the Deployment and try the creation again.
 	case errors.IsAlreadyExists(err):
 		alreadyExists = true
 		rs, rsErr := dc.rsLister.ReplicaSets(newRS.Namespace).Get(newRS.Name)
 		if rsErr != nil {
 			return nil, rsErr
 		}
+
+		controllerRef := metav1.GetControllerOf(rs)
+		if controllerRef != nil && controllerRef.UID == d.UID && deploymentutil.EqualIgnoreHash(&d.Spec.Template, &rs.Spec.Template) {
+			createdRS = rs
+			err = nil
+			break
+		}
+
```

Comment:
What if `rs` is an orphan that can be adopted by the Deployment as its new RS? This could happen when the user create an RS first and then create a Deployment to adopt it. 

If `rs` can be adopted by the Deployment, we should make the decision in the next control loop (after it's adopted) before bumping collision count. Otherwise this `rs` will have an unnecessary scale down.

Reply:
If the RS is an orphan that can be adopted, ControllerRefManager will adopt it on the next sync, before Deployment decides whether to create a new RS.

After we bump the collision count, we requeue rather than immediately create a new RS. So if the RS can be adopted in the next sync, it shouldn't matter that we bumped the collision count unnecessarily.

Reply:
Worth adding a test for this case.
1. Create a Deployment, wait for its new RS to be created 
2. Remove the Deployment and orphan the RS
3. Create the Deployment again, check that the Deployment only manages one RS after the RS is adopted

Reply:
Yes, there should be another test for further testing against collision avoidance mechanism, which will be addressed in another PR. The PR has changed to xref the issue.

Reply:
We have an e2e test for that, and I think it makes sense there; we expect the user may intentionally delete the Deployment without cascading and recreate it, so it's not an edge case.

https://github.com/kubernetes/kubernetes/blob/bf73cdf4e7f5e6bd81438ba84b1b60896706459a/test/e2e/apps/deployment.go#L749-L786

However, it doesn't appear that the test ensures no extra RS was created, so that could be a good improvement to make to the e2e test.

Reply:
Great! I will link this test to the issue, and work on it.

Reply:
Nice!